qiskit/algorithms/optimizers/steppable_optimizer.py

```diff
@@ -0,0 +1,213 @@
+from abc import abstractmethod, ABC
+from dataclasses import dataclass, field
+from typing import Dict, Any, Union, Callable, Optional, Tuple, List
+
+from .optimizer import Optimizer, POINT, OptimizerResult
+import numpy as np
+
+CALLBACK = Callable[[int, np.ndarray, float, float], None]
+
+
+@dataclass
+class AskObject(ABC):
+    """Base class for return type of method SteppableOptimizer.ask()"""
+
+    x_fun: Union[POINT, List[POINT]]
+    x_jac: Union[POINT, List[POINT]]
+
+
+@dataclass
+class TellObject(ABC):
+    """Base class for argument type of method SteppableOptimizer.tell()"""
+
+    eval_fun: Union[float, List[float]]
+    eval_jac: Union[POINT, List[POINT]]
+
+
+@dataclass
+class OptimizerState:
+    """
+    Base class representing the state of the optmiizer.
+    Any variable that changes during the optimization process and is needed for the next step
+    of optimization should be stored in this dataclass.
+    """
+
+    x: POINT  # pylint: disable=invalid-name
+    fun: Optional[Callable[[POINT], float]]  # Make optional
+    jac: Optional[Callable[[POINT], POINT]]
+    nfev: Optional[int]
+    njev: Optional[int]
+    nit: Optional[int]
+
+
+class SteppableOptimizer(Optimizer):
+    """
+    Base class for a steppable optimizer.
+
+    This family of optimizers will be using the ask and tell interface.
+    When using this interface the user has to call the function ask() in order to get information about how to evaluate the fucntion
+    (we are asking the optimizer about how to do the evaluation). This information will be mostly about at what point should we evaluate
+    the function next, but depending on the optimizer it can also be about whether we should evaluate the function itself or its gradient.
+
+    Once the function has been evaluated, the user calls the method tell() to tell the optimizer what has been the result of the function evaluation.
+    The optimizer then updates its state accordingly and the user can decide whether to stop the optimization process or to repeat a step.
+
+    This interface is more customizable, and allows the user to have full control over the evaluation of the function.
+
+    For example:
+    .. code-block::python
+
```

Comment:
Could you include all the imports needed for this to run so that users can just copy+paste this snippet? 🙂